KAFKA-2752: Add VerifiableSource/Sink connectors and rolling bounce Copycat system tests.

ewencp opened the pull request.

ewencp committed:
KAFKA-2752: Add VerifiableSource/Sink connectors and rolling bounce Copycat system tests.

ewencp commented:
This requires https://github.com/apache/kafka/pull/431 before this test will run reliably.

You'll probably also notice a bunch of logging related changes. These were some changes that were helpful in tracking down issues in the 4 or 5 JIRAs that came out of writing this test, so I think they're worth keeping.

wushujames commented on copycat/tools/src/main/java/org/apache/kafka/copycat/tools/VerifiableSourceTask.java:
```diff
@@ -0,0 +1,124 @@
+import java.util.Properties;
+
+/**
+ * A connector primarily intended for system tests. The connector simply generates as many tasks as requested. The
```
Nice. This is a cool way test distributed mode.

wushujames commented on tests/kafkatest/tests/copycat_distributed_test.py:
```diff
@@ -76,19 +79,103 @@ def test_file_source_and_sink(self, converter="org.apache.kafka.copycat.json.Jso
 
-    def validate_output(self, input):
+
+    def test_clean_bounce(self):
```
Nice!

guozhangwang commented on clients/src/main/java/org/apache/kafka/common/utils/ThroughputThrottler.java:
```diff
@@ -96,12 +101,19 @@ public void throttle() {
             try {
-                Thread.sleep(sleepMs, (int) sleepNs);
+                synchronized (this) {
+                    long elapsed = currentTimeNs - sleepStartNs;
```
Is this correct? currentTimeNs is never updated and hence always the same as sleepStartNs.

ewencp replied:
No, it was not correct. Good catch! I've fixed it and validated correctness by running ProducerPerformance manually.

guozhangwang commented on tests/kafkatest/tests/copycat_distributed_test.py:
```diff
@@ -76,19 +79,103 @@ def test_file_source_and_sink(self, converter="org.apache.kafka.copycat.json.Jso
+            missing_sink_seqnos = sorted(set(range(sink_seqno_max)).difference(set(sink_seqnos)))
+            duplicate_sink_seqnos = sorted([seqno for seqno,count in sink_seqno_counts.iteritems() if count > 1])
+
+            if missing_sink_seqnos:
```
Shall we compare the set from the source and sink to make sure they are the same as well?

benstopford replied:
I think it should be valid as is, although it obviously wouldn't hurt.

ewencp replied:
@guozhangwang @benstopford We can't guarantee they are exactly the same. The problem is deciding when it is safe to stop the sink connector. Currently we just kill it quickly after killing the source, so we can't be sure all the data generated by the source has been processed.

We could do something like tail the output of all the nodes running copycat workers, waiting until the last message produced is flushed by a sync. I wasn't convinced this was worth a) the more complicated test code and b) the fact that you then have to stream in realtime all the output of the workers.

benstopford commented on tests/kafkatest/tests/copycat_distributed_test.py:
```diff
@@ -68,7 +71,7 @@ def test_file_source_and_sink(self, converter="org.apache.kafka.copycat.json.Jso
         for node in self.cc.nodes:
             node.account.ssh("echo -e -n " + repr(self.FIRST_INPUTS) + " >> " + self.INPUT_FILE)
-        wait_until(lambda: self.validate_output(self.FIRST_INPUT_LIST), timeout_sec=120, err_msg="Data added to input file was not seen in the output file in a reasonable amount of time.")
+        wait_until(lambda: self._validate_file_output(self.FIRST_INPUT_LIST), timeout_sec=120, err_msg="Data added to input file was not seen in the output file in a reasonable amount of time.")
```
minor... but 2mins seems a long time to give the process to start.

ewencp replied:
It is. This could probably be dropped to be lower now, though I'd rather keep them rather liberal to avoid an spurious errors just due to timing issues.

The problem with Copycat startup in particular is that in distributed mode it may need to join the worker group which has a session timeout of 30s and any sink tasks that start also have a potential 30s timeout. This is also related to https://issues.apache.org/jira/browse/KAFKA-2713. Now that leave group has been implemented, this _should_ be ok at much lower value since this class assumes clean bouncing of nodes.

For now, I think I'd like to leave the timeouts relatively high and avoid spurious failures that are just due to being too conservative. I think sometime after the 0.9.0.0 release, we're going to want to do a general cleanup of things like timeouts and get a better handle on system test runtimes.

benstopford commented:
This looks reasonable to me

ewencp committed:
Address review comments.

gwenshap commented:
Just note that the PR title is wrong - its KAFKA-2752 (not 2572).

We don't want to close the wrong JIRA accidentally :)

ewencp committed:
Add debugging support by collecting a dump of the topic in case of failure and logging a bit more info during the test run.

ewencp commented:
@gwenshap Good catch, fixed in the commit msg and the PR title.

@guozhangwang This is effectively complete wrt the actual test (and I added a bit more code to dump the contents of the topic in the case of failure, which can aid in debugging). However, we'll still hit failures somtimes. On the source side, we can see duplicates sometimes, sort of due to https://issues.apache.org/jira/browse/KAFKA-2713 where a sink task can block during join for long enough that it hits the worker group timeout and the rest of the group moves on, but the source task then processes data even after its out of the group (and can commit offsets since it thinks it still owns that data).

I've seen missing sink data, but haven't tracked down the issue yet. I haven't manage to catch it with good enough logging + a dump of the topic yet.

We can also see some unhandled exceptions currently because of wakeup exceptions during consumer.close(), an issue that was only introduced recently. (I'm coordinating w/ @hachikuji to sort that issue out).

So, probably a question for the community more generally and @guozhangwang and @gwenshap immediately: how do we want to handle tests that we're confident are setup properly and useful, but are going to fail while some existing issues are sorted out? I kind of don't want this to sit in limbo (I already did that with the first 5 patches that writing this test resulted in....). On the other hand, committing failing tests is not helpful. We have an `@ignore` decorator, but that disables running the test entirely. @granders and I were discussing an `@unstable` annotation, but I'm not sure how quickly we can get that because making the semantics useful (i.e. actually notifies you of failures, but in a nice way if only unstable tests fail, doesn't require proactively checking test results, doesn't require a bunch of 1-liner commits to remove unstable annotations once the test is stabilized, etc).

Thoughts about what to do with this patch in the mean time until those issues are addressed?

guozhangwang commented:
@ewencp It will be a bug if the kick-out-of-group consumer can still commit offsets: their commit should be rejected with ILLEGAL_GENERATION. If you see this happening we should investigate this case asap I think.

About unit testing: I would personally prefer only adding test cases that are passing to the code base, and piggy-back tests that are failing due to some issues to the fix of that issue. But I also understands keeping track of those tests is sort of a pain: more painful it is, more motivated are we to resolve the dandling problem. So probably we should just eat the bullet and fix the related existing issue right away?

ewencp commented:
@guozhangwang The issue is with source connectors, which have to handle offset commits outside the normal consumer group offset commit functionality since they do not use topic partitions or integer offsets. So this issue is definitely due to the fact that we have not generalized a mechanism for having the group coordinator manage these types of write/commit operations in a more general manner. I thought about routing all these writes through the leader, but that only reduces, not removes, the issue since the lagging node could be the leader.

Re: tests, yes, I am looking to fix them immediately, so in this case we can defer for a while at least. I guess part of my point is that in general system tests are a lot more finicky than unit tests, so even if it's been reviewed and passes regularly locally, we can still see cases where it fails (especially for new functionality like this!). So even when we think the test is stable, we still might want a mechanism to put it into "trial mode" for awhile. But that's a longer term discussion, we don't have to address it right now.

ewencp committed:
Increase log level to DEBUG for Copycat services since this level will be required to debug the distributed test failures.

ewencp committed:
KAFKA-2713: Run task start and stop methods in worker threads so they execute in parallel and cannot block the herder thread.

ewencp committed:
Merge remote-tracking branch 'origin/trunk' into kafka-2752-copycat-clean-bounce-test

ewencp committed:
Merge remote-tracking branch 'origin/trunk' into kafka-2752-copycat-clean-bounce-test

ewencp committed:
Merge remote-tracking branch 'origin/trunk' into kafka-2752-copycat-clean-bounce-test

ewencp committed:
Fix connect:tools build settings.

ewencp committed:
Merge remote-tracking branch 'origin/trunk' into kafka-2752-copycat-clean-bounce-test

ewencp commented:
I think I've got this stabilized via a number of other bug fixes that are now committed. Two issues remain: #476 and #480. When those are merged, I can update this branch and we can verify one final time that this is now stable.

ewencp committed:
Merge remote-tracking branch 'origin/trunk' into kafka-2752-copycat-clean-bounce-test

ewencp committed:
Fix leftover copycat reference.

ewencp commented:
@gwenshap @guozhangwang I've updated this with trunk and it's now stable as far as I can tell (passed many times in a row).

guozhangwang commented:
One minor question, otherwise LGTM!

granders commented on tests/kafkatest/services/connect.py:
```diff
@@ -37,7 +53,7 @@ def __init__(self, context, num_nodes, kafka, files):
         """Return process ids for Kafka Connect processes."""
         try:
-            return [pid for pid in node.account.ssh_capture("cat /mnt/connect.pid", callback=int)]
+            return [pid for pid in node.account.ssh_capture("cat " + self.PID_FILE, callback=int)]
```
pid files are good :) However `clean_node` probably should not rely on the pid file

ewencp replied:
Fair point, replaced with a `kill_process` version.

ewencp committed:
Replace clean_node implementation with more aggressive implementation that kills all connect processes rather than using pid files.

ewencp committed:
Fix typo in log4j settings file.

granthenke commented:
See #492 for checkstyle quick fix
